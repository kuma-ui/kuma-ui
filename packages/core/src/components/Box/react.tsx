import { StyledProps, PseudoProps } from "@kuma-ui/system";
import React, { ReactNode } from "react";
import { As, ComponentWithAs, MergeWithAs, PropsOf } from "../types";

type BoxProps = StyledProps &
  Partial<PseudoProps> & {
    children?: ReactNode;
  };

type BoxComponent<T extends As = "div"> = ComponentWithAs<T, BoxProps>;

/**
 * Box is the most abstract component in Kuma UI, providing a base upon which all other components are built.
 * It renders a div element by default, and can accept any system style properties for extensive customization.
 *
 * @see â€” Further documentation will be available in the future.
 */
const Box: BoxComponent = <T extends As = "div">({
  as: Component = "div",
  children,
  ...props
}: MergeWithAs<PropsOf<T>, BoxProps>) =>
  React.createElement(Component, props, children);

export { Box, type BoxComponent, BoxProps };
