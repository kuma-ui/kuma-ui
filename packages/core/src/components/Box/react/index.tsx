import React from "react";
import type { BoxComponent, BoxProps } from "./types";
import { DynamicBox } from "./DynamicBox";
import { StaticBox } from "./StaticBox";
import { hasDynamicProps } from "./utils";

/**
 * Box is the most abstract component in Kuma UI, providing a base upon which all other components are built.
 * It renders a div element by default, and can accept any system style properties for extensive customization.
 *
 * @see â€” http://kuma-ui.com/docs/Components/Box
 */
const Box: BoxComponent = ({ children, ...props }) => {
  if (hasDynamicProps(props)) {
    return React.createElement(DynamicBox, props, children);
  }

  return React.createElement(StaticBox, props, children);
};

export { Box, type BoxComponent, BoxProps };
