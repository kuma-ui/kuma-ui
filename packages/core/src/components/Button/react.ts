import { StyledProps, PseudoProps } from "@kuma-ui/system";
import React, { ReactNode } from "react";
import { As, ComponentWithAs, MergeWithAs, PropsOf } from "../types";

type ButtonProps = StyledProps &
  Partial<PseudoProps> & {
    children?: ReactNode;
  };

type ButtonComponent<T extends As = "button"> = ComponentWithAs<T, ButtonProps>;

/**
 * Button is a clickable component used to trigger actions or events in a Kuma UI application.
 *
 * @see â€” Further documentation will be available in the future.
 */
const Button: ButtonComponent = <T extends As = "button">({
  as: Component = "button",
  children,
  ...props
}: MergeWithAs<PropsOf<T>, ButtonProps>) =>
  React.createElement(Component, props, children);

export { Button, type ButtonComponent, ButtonProps };
