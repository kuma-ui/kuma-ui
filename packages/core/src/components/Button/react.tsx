import { StyledProps, PseudoProps } from "@kuma-ui/system";
import React, { ReactNode } from "react";
import {
  As,
  ComponentWithAs,
  MergeWithAs,
  PropsOf,
  ComponentProps,
} from "../types";
import { Box } from "../Box";
import { theme } from "@kuma-ui/sheet";
import { defaultButtonTag } from "./handler";

type ButtonProps = ComponentProps<"Button">;

type ButtonComponent<T extends As = "button"> = ComponentWithAs<T, ButtonProps>;

/**
 * Button is a clickable component used to trigger actions or events in a Kuma UI application.
 *
 * @see â€” http://kuma-ui.com/docs/Components/Button
 */
const Button: ButtonComponent = <T extends As = "button">({
  as: Component = defaultButtonTag,
  children,
  ...props
}: MergeWithAs<PropsOf<T>, ButtonProps>) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
  const variant = props.variant
    ? theme.getVariants("Button")?.variants?.[props.variant]
    : {};
  return (
    <Box
      as={Component}
      {...variant}
      {...props}
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
      children={children}
      IS_KUMA_DEFAULT
    />
  );
};

export { Button, type ButtonComponent, ButtonProps };
