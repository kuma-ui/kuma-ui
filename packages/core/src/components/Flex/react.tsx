import { StyledProps, PseudoProps } from "@kuma-ui/system";
import { theme } from "@kuma-ui/sheet";
import React, { ReactNode } from "react";
import {
  As,
  ComponentWithAs,
  MergeWithAs,
  PropsOf,
  ComponentProps,
} from "../types";
import { Box } from "../Box";
import { defaultFlexTag } from "./handler";

type FlexProps = ComponentProps<"Flex">;

type FlexComponent<T extends As = "div"> = ComponentWithAs<T, FlexProps>;

/**
 * Flex is a Box component with 'display: flex' by default. It's used to create flexible layouts in a Kuma UI application.
 * Like Box, it can accept any system style properties for extensive customization.
 *
 * @see â€” http://kuma-ui.com/docs/Components/Flex
 */
const Flex: FlexComponent = <T extends As = "div">({
  as: Component = defaultFlexTag,
  children,
  ...props
}: MergeWithAs<PropsOf<T>, FlexProps>) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
  const variant = props.variant
    ? theme.getVariants("Flex")?.variants?.[props.variant]
    : {};
  return (
    <Box
      as={Component}
      {...variant}
      {...props}
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
      children={children}
      IS_KUMA_DEFAULT
    />
  );
};

export { Flex, type FlexComponent, FlexProps };
