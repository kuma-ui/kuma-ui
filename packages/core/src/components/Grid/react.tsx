import { theme } from "@kuma-ui/sheet";
import React, { ReactNode } from "react";
import {
  As,
  ComponentWithAs,
  MergeWithAs,
  PropsOf,
  ComponentProps,
} from "../types";
import { Box } from "../Box";
import { defaultGridTag } from "./handler";

type GridProps = ComponentProps<"Grid">;

type GridComponent<T extends As = "div"> = ComponentWithAs<T, GridProps>;

/**
 * Grid is a powerful component that allows for 2-dimensional layouts and alignment of child components.
 *
 * @see â€” http://kuma-ui.com/docs/Components/Grid
 */
const Grid: GridComponent = <T extends As = "div">({
  as: Component = defaultGridTag,
  children,
  ...props
}: MergeWithAs<PropsOf<T>, GridProps>) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
  const variant = props.variant
    ? theme.getVariants("Grid")?.variants?.[props.variant]
    : {};
  return (
    <Box
      as={Component}
      {...variant}
      {...props}
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
      children={children}
      IS_KUMA_DEFAULT
    />
  );
};

export { Grid, type GridComponent, GridProps };
