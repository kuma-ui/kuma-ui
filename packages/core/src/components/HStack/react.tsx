import { StyledProps, PseudoProps } from "@kuma-ui/system";
import { theme } from "@kuma-ui/sheet";
import React, { ReactNode } from "react";
import {
  As,
  ComponentWithAs,
  MergeWithAs,
  PropsOf,
  ComponentProps,
} from "../types";
import { Box } from "../Box";
import { defaultHStackTag } from "./handler";

type HStackProps = ComponentProps<"HStack">;

type HStackComponent<T extends As = "div"> = ComponentWithAs<T, HStackProps>;

/**
 * HStack is a Box component with `display="flex"` and `flex-direction: row` set by default. It makes it easier to layout and align child components in a horizontal direction. It is a fundamental layout component in Kuma UI.
 *
 * @see â€” http://kuma-ui.com/docs/Components/HStack
 */
const HStack: HStackComponent = <T extends As = "div">({
  as: Component = defaultHStackTag,
  children,
  ...props
}: MergeWithAs<PropsOf<T>, HStackProps>) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
  const variant = props.variant
    ? theme.getVariants("HStack")?.variants?.[props.variant]
    : {};
  return (
    <Box
      as={Component}
      {...variant}
      {...props}
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
      children={children}
      IS_KUMA_DEFAULT
    />
  );
};

export { HStack, type HStackComponent, HStackProps };
