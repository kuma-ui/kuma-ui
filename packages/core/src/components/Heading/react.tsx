import { StyledProps, PseudoProps } from "@kuma-ui/system";
import React, { ReactNode } from "react";
import {
  As,
  ComponentWithAs,
  MergeWithAs,
  PropsOf,
  ComponentProps,
} from "../types";
import { Box } from "../Box";
import { theme } from "@kuma-ui/sheet";
import { defaultHeadingTag } from "./handler";

type HeadingProps = ComponentProps<"Heading">;

type HeadingComponent<
  T extends "h1" | "h2" | "h3" | "h4" | "h5" | "h6" = "h1"
> = ComponentWithAs<T, HeadingProps>;

/**
 * Used to render semantic HTML heading elements in a Kuma UI application.
 * By default, renders as h1 which can be customized with the 'as' prop.
 *
 * @see â€” http://kuma-ui.com/docs/Components/Heading
 */
const Heading: HeadingComponent = <
  T extends "h1" | "h2" | "h3" | "h4" | "h5" | "h6" = "h1"
>({
  as: Component = defaultHeadingTag,
  children,
  ...props
}: MergeWithAs<PropsOf<T>, HeadingProps>) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
  const variant = props.variant
    ? // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any -- FIXME
      theme.getVariants("Heading")?.variants?.[props.variant as any]
    : {};
  return (
    <Box
      as={Component}
      {...variant}
      {...props}
      children={children}
      IS_KUMA_DEFAULT
    />
  );
};

export { Heading, type HeadingComponent, HeadingProps };
