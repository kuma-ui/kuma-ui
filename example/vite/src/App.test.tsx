import App from "./App";
import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

describe("App", () => {
  it("renders unchanged", () => {
    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });
});
