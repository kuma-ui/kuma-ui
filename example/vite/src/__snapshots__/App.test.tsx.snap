// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders unchanged 1`] = `
<DocumentFragment>
  .kuma-0 {
  padding: 2px;
  color: red;
}

.kuma-1 {
  color: red;
}

.kuma-2 {
  color: blue;
}

.kuma-3 {
  padding: 10px 12px;
  font-size: 16px;
}

@media (min-width:400px) {
  .kuma-0 {
    padding: 4px;
  }
}

.kuma-0:hover {
  color: black;
}

  <div>
    <div
      class="kuma-0"
    >
      hello
    </div>
    <p
      class="kuma-2 kuma-1"
    >
      world
    </p>
    <button
      class="kuma-3"
      disabled=""
    >
      button
    </button>
  </div>
</DocumentFragment>
`;
